File: apps/builder/FieldTakHub.Builder/Services/UpdateService.cs
using System.IO;
using System.Diagnostics;
using System.Net.Http;
using System.Net.Http.Headers;
using System.Security.Cryptography;
using System.Text.Json;

namespace FieldTakHub.Builder.Services;

public sealed record BuilderUpdateInfo(string Version, string AssetName, string AssetUrl, string Sha256, string ReleasePage, string Channel);

public sealed class UpdateService
{
    private readonly HttpClient _http;
    private readonly string _repository;
    private readonly string _channel;

    public UpdateService()
    {
        var cfgPath = Path.Combine(AppContext.BaseDirectory, "update-config.json");
        var repo = "Qbcol/TAKFieldHub"; var channel = "rc";
        if (File.Exists(cfgPath))
        {
            using var doc = JsonDocument.Parse(File.ReadAllText(cfgPath));
            repo = doc.RootElement.TryGetProperty("repository", out var r) ? r.GetString() ?? repo : repo;
            channel = doc.RootElement.TryGetProperty("channel", out var c) ? c.GetString() ?? channel : channel;
        }
        _repository = repo; _channel = channel;
        _http = new HttpClient(new HttpClientHandler { AllowAutoRedirect = false }) { Timeout = TimeSpan.FromSeconds(30) };
        _http.DefaultRequestHeaders.UserAgent.Add(new ProductInfoHeaderValue("FieldTAKHubBuilder", "2.2.0-rc1"));
        _http.DefaultRequestHeaders.Accept.Add(new MediaTypeWithQualityHeaderValue("application/vnd.github+json"));
    }

    public async Task<BuilderUpdateInfo?> CheckAsync(CancellationToken ct = default)
    {
        if (string.IsNullOrWhiteSpace(_repository) || !_repository.Contains('/')) return null;
        var releasesJson = await GetHttpsTextAsync($"https://api.github.com/repos/{_repository}/releases?per_page=20", ct);
        using var releases = JsonDocument.Parse(releasesJson);
        JsonElement? chosen = null;
        foreach (var release in releases.RootElement.EnumerateArray())
        {
            if (release.TryGetProperty("draft", out var draft) && draft.GetBoolean()) continue;
            var prerelease = release.TryGetProperty("prerelease", out var pre) && pre.GetBoolean();
            if ((_channel.Equals("rc", StringComparison.OrdinalIgnoreCase) && prerelease) || (_channel.Equals("stable", StringComparison.OrdinalIgnoreCase) && !prerelease)) { chosen = release.Clone(); break; }
        }
        if (chosen is null) return null;
        var selected = chosen.Value; var assets = selected.GetProperty("assets").EnumerateArray().Select(a=>a.Clone()).ToList();
        var manifestAsset = assets.FirstOrDefault(a => a.GetProperty("name").GetString() == "fieldtak-release.json");
        if (manifestAsset.ValueKind == JsonValueKind.Undefined) return null;
        var manifestUrl = manifestAsset.GetProperty("browser_download_url").GetString();
        if (string.IsNullOrWhiteSpace(manifestUrl)) return null;
        using var manifest = JsonDocument.Parse(await GetHttpsTextAsync(manifestUrl, ct));
        if (!string.Equals(manifest.RootElement.GetProperty("channel").GetString(), _channel, StringComparison.OrdinalIgnoreCase)) return null;
        var builder = manifest.RootElement.GetProperty("builder");
        var version = builder.GetProperty("version").GetString() ?? "";
        if (CompareVersions(version, "2.2.0-rc1") <= 0) return null;
        var assetName = builder.GetProperty("asset").GetString() ?? "";
        var sha = builder.GetProperty("sha256").GetString()?.ToLowerInvariant() ?? "";
        var match = assets.FirstOrDefault(a => a.GetProperty("name").GetString() == assetName);
        var url = match.ValueKind == JsonValueKind.Undefined ? null : match.GetProperty("browser_download_url").GetString();
        if (string.IsNullOrWhiteSpace(url) || sha.Length != 64) return null;
        return new(version, assetName, url, sha, selected.GetProperty("html_url").GetString() ?? "", _channel);
    }

    public async Task<string> DownloadAsync(BuilderUpdateInfo update, IProgress<double>? progress = null, CancellationToken ct = default)
    {
        var target = Path.Combine(Path.GetTempPath(), update.AssetName);
        using var response = await SendHttpsAsync(update.AssetUrl, ct);
        response.EnsureSuccessStatusCode();
        var total = response.Content.Headers.ContentLength;
        await using var input = await response.Content.ReadAsStreamAsync(ct);
        await using var output = File.Create(target);
        var buffer = new byte[128 * 1024]; long current = 0;
        while (true)
        {
            var n = await input.ReadAsync(buffer, ct); if (n <= 0) break;
            await output.WriteAsync(buffer.AsMemory(0, n), ct); current += n;
            if (total is > 0) progress?.Report((double)current / total.Value);
        }
        await using var hashStream = File.OpenRead(target);
        var actual = Convert.ToHexString(await SHA256.HashDataAsync(hashStream, ct)).ToLowerInvariant();
        if (!actual.Equals(update.Sha256, StringComparison.OrdinalIgnoreCase)) throw new InvalidDataException("FTH-UPD-104: downloaded Builder SHA-256 mismatch.");
        return target;
    }

    public static void ShowInExplorer(string file) => Process.Start(new ProcessStartInfo("explorer.exe", $"/select,\"{file}\"") { UseShellExecute = true });

    private async Task<string> GetHttpsTextAsync(string url, CancellationToken ct)
    {
        using var response = await SendHttpsAsync(url, ct); response.EnsureSuccessStatusCode(); return await response.Content.ReadAsStringAsync(ct);
    }

    private async Task<HttpResponseMessage> SendHttpsAsync(string initial, CancellationToken ct)
    {
        var current = RequireHttps(initial);
        for (var i = 0; i < 6; i++)
        {
            var response = await _http.GetAsync(current, HttpCompletionOption.ResponseHeadersRead, ct);
            if ((int)response.StatusCode is >= 300 and < 400)
            {
                var location = response.Headers.Location ?? throw new InvalidDataException("FTH-UPD-101: redirect without Location.");
                current = RequireHttps(location.IsAbsoluteUri ? location.ToString() : new Uri(new Uri(current), location).ToString());
                response.Dispose(); continue;
            }
            return response;
        }
        throw new InvalidDataException("FTH-UPD-102: too many redirects.");
    }

    private static string RequireHttps(string url)
    {
        var uri = new Uri(url, UriKind.Absolute);
        if (!uri.Scheme.Equals(Uri.UriSchemeHttps, StringComparison.OrdinalIgnoreCase)) throw new InvalidDataException("FTH-UPD-103: update URL must use HTTPS.");
        return uri.ToString();
    }

    private static int CompareVersions(string a, string b)
    {
        static int[] V(string x) => x.Split('-', 2)[0].Split('.').Select(v => int.TryParse(v, out var n) ? n : 0).Concat(Enumerable.Repeat(0, 4)).Take(4).ToArray();
        var av=V(a); var bv=V(b); for(var i=0;i<4;i++){var c=av[i].CompareTo(bv[i]);if(c!=0)return c;} return string.Compare(a,b,StringComparison.OrdinalIgnoreCase);
    }
}

File: apps/android/app/src/main/java/org/fieldtak/hub/update/UpdateService.kt
package org.fieldtak.hub.update

import android.content.Context
import org.fieldtak.hub.BuildConfig
import org.fieldtak.hub.security.UrlPolicy
import org.fieldtak.hub.util.VersionUtil
import org.json.JSONArray
import org.json.JSONObject
import java.io.File
import java.net.HttpURLConnection
import java.net.URL
import java.security.MessageDigest

class UpdateService(private val context:Context) {
  private val userAgent = "FieldTAKHub/${BuildConfig.VERSION_NAME}"

  fun check():AppUpdateInfo? {
    val repo=BuildConfig.UPDATE_REPOSITORY.trim()
    if(repo.isBlank() || !repo.contains('/')) return null
    val releases=JSONArray(getText("https://api.github.com/repos/$repo/releases?per_page=20"))
    var selected:JSONObject?=null
    for(i in 0 until releases.length()){
      val r=releases.getJSONObject(i)
      if(r.optBoolean("draft",false)) continue
      val prerelease=r.optBoolean("prerelease",false)
      if((BuildConfig.RELEASE_CHANNEL=="rc" && prerelease) || (BuildConfig.RELEASE_CHANNEL=="stable" && !prerelease)){ selected=r; break }
    }
    val release=selected ?: return null
    val assets=release.optJSONArray("assets") ?: return null
    var manifestUrl:String?=null
    val downloadByName=linkedMapOf<String,String>()
    for(i in 0 until assets.length()){
      val a=assets.getJSONObject(i); val name=a.optString("name"); val url=a.optString("browser_download_url")
      if(name.isNotBlank() && url.isNotBlank()) downloadByName[name]=url
      if(name=="fieldtak-release.json") manifestUrl=url
    }
    val manifest=JSONObject(getText(manifestUrl ?: return null))
    if(manifest.optString("channel") != BuildConfig.RELEASE_CHANNEL) return null
    val hub=manifest.optJSONObject("hub") ?: return null
    val version=hub.optString("version"); val versionCode=hub.optLong("versionCode",0)
    val asset=hub.optString("asset"); val sha=hub.optString("sha256").lowercase()
    if(version.isBlank() || asset.isBlank() || sha.length!=64) return null
    if(versionCode <= BuildConfig.VERSION_CODE && VersionUtil.compare(version,BuildConfig.VERSION_NAME)<=0) return null
    val assetUrl=downloadByName[asset] ?: return null
    return AppUpdateInfo(version,versionCode,asset,assetUrl,sha,release.optString("html_url"),manifest.optString("channel"))
  }

  fun download(update:AppUpdateInfo, progress:(Long,Long?)->Unit={_,_->}):File {
    val dir=File(context.filesDir,"updates").apply{mkdirs()}
    val part=File(dir,update.assetName+".part"); val done=File(dir,update.assetName)
    if(done.isFile && sha256(done).equals(update.sha256,true)) return done
    val conn=openHttps(update.assetUrl)
    try {
      require(conn.responseCode in 200..299){"FTH-UPD-003: HTTP ${conn.responseCode}"}
      val total=conn.getHeaderFieldLong("Content-Length",-1L).takeIf{it>=0}
      conn.inputStream.use { input -> part.outputStream().use { out ->
        val buf=ByteArray(128*1024); var current=0L
        while(true){val n=input.read(buf);if(n<0)break;out.write(buf,0,n);current+=n;progress(current,total)}
      }}
    } finally { conn.disconnect() }
    require(sha256(part).equals(update.sha256,true)){"FTH-UPD-004: downloaded APK SHA-256 mismatch"}
    if(done.exists()) done.delete(); require(part.renameTo(done)){"FTH-UPD-005: cannot finalize update"}; return done
  }

  private fun getText(url:String):String {
    val c=openHttps(url)
    return try { require(c.responseCode in 200..299){"FTH-UPD-006: HTTP ${c.responseCode}"}; c.inputStream.bufferedReader().use{it.readText()} }
    finally { c.disconnect() }
  }

  private fun openHttps(initial:String):HttpURLConnection {
    var current=UrlPolicy.requireHttps(initial)
    repeat(6) {
      val c=(URL(current).openConnection() as HttpURLConnection).apply {
        connectTimeout=12_000; readTimeout=30_000; instanceFollowRedirects=false; requestMethod="GET"
        setRequestProperty("User-Agent",userAgent); setRequestProperty("Accept","application/vnd.github+json")
      }
      val code=c.responseCode
      if(code in 300..399){ val next=c.getHeaderField("Location") ?: error("FTH-UPD-007: redirect without Location"); c.disconnect(); current=UrlPolicy.requireHttps(URL(URL(current),next).toString()) }
      else return c
    }
    error("FTH-UPD-008: too many redirects")
  }

  private fun sha256(file:File):String {
    val md=MessageDigest.getInstance("SHA-256")
    file.inputStream().buffered().use{input->val b=ByteArray(128*1024);while(true){val n=input.read(b);if(n<0)break;md.update(b,0,n)}}
    return md.digest().joinToString(""){"%02x".format(it)}
  }
}
